cpp/src/unary/unary_ops.cuh

```diff
@@ -0,0 +1,133 @@
+        for (int i=start; i<size; i+=step) {
+            if ( gdf_is_valid(valid, i) )
+                results[i] = functor.apply(data[i]);
+        }
+    } else {        // no valid mask
+        for (int i=start; i<size; i+=step) {
+            results[i] = functor.apply(data[i]);
+        }
+    }
+}
+
+template<typename T, typename Tout, typename F>
+struct Launcher {
+    static
+    gdf_error launch(gdf_column *input, gdf_column *output) {
+
+        // Return immediately for empty inputs
+        if((0==input->size))
+        {
+          return GDF_SUCCESS;
+        }
+
+        /* check for size of the columns */
+        if (input->size != output->size) {
+            return GDF_COLUMN_SIZE_MISMATCH;
+        }
+
+        // find optimal blocksize
+        int mingridsize, blocksize;
+        CUDA_TRY(
+            cudaOccupancyMaxPotentialBlockSize(&mingridsize, &blocksize,
+                                               gpu_op_kernel<T, Tout, F>)
+        );
+        // find needed gridsize
+        int neededgridsize = (input->size + blocksize - 1) / blocksize;
+        int gridsize = std::min(neededgridsize, mingridsize);
+
+        F functor;
+        gpu_op_kernel<<<gridsize, blocksize>>>(
+            // input
+            (const T*)input->data, input->valid, input->size,
+            // output
+            (Tout*)output->data,
+            // action
+            functor
+        );
+
+        CUDA_CHECK_LAST();
+        return GDF_SUCCESS;
+    }
+};
+
+inline void handleChecksAndValidity(gdf_column *input, gdf_column *output) {
+    // Check for null pointers in input
+    CUDF_EXPECTS((input != nullptr), "Pointer to input column is null");
+    CUDF_EXPECTS((output != nullptr), "Pointer to output column is null");
+
+    // Check for null data pointer
+    CUDF_EXPECTS((input->data != nullptr),
+        "Pointer to data in input column is null");
+    CUDF_EXPECTS((output->data != nullptr),
+        "Pointer to data in output column is null");
+
+    // Check if input has valid mask if null_count > 0
+    CUDF_EXPECTS((input->null_count == 0 || input->valid != nullptr),
+        "Pointer to input column's valid mask is null but null count > 0");
+
+    if (input->valid == nullptr) {
+        if (output->valid == nullptr) {
+            // if input column has no mask, then output column is allowed to have no mask
+            output->null_count = 0;
+        }
+        else { // output->valid != nullptr
+            CUDA_TRY( cudaMemset(output->valid, 0xff,
+                                gdf_num_bitmask_elements( input->size )) );
+            output->null_count = 0;
+        }
+    }
+    else { // input->valid != nullptr
+        CUDF_EXPECTS((output->valid != nullptr),
+            "Input column has valid mask but output column does not");
+
+        // Validity mask transfer
+        CUDA_TRY( cudaMemcpy(output->valid, input->valid,
+                             gdf_num_bitmask_elements( input->size ),
```

Comment:
@jrhemstad Should this be `gdf_valid_allocation_size`?

Reply:
Seems like it shouldn't matter since you are just copying the useful bytes of the valid mask. Anything that you miss with this will just be padding. Agree?

Reply:
alright. I'll leave it be